### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by Felipe0542**
Módulo 4:

Con la misma estructura que en el anterior, entrego el notebook, dos scripts y el markdown.

Hice las entregas 3 y 4 al mismo tiempo, en esta he incorporado cambios que me comentas en la corrección. Sigo la estructura de ir probando modelos (en este caso Random Forest, Gradiente y Neural Network) primero sin las variables categóricas y luego las prueblo haciendo rodinal y one hot encoding.

El modelo final que selecciono es un Random Forest one hot encoding (aunque limitado para no elevar mucho los tiempos de entrenar el modelo).

Para todos los modelos he ido probando varios hiperparámetros. El baseline he apartado el lineal y he tenido en cuenta una regresión logística con una regularización l2.

Como algunos gráficos ya los tenía y habían sido un poco cargados de hacer (mucho tiempo) los he mantenido y añadido partes nuevas que comparan directamente mi nuevo baseline con el mejor modelo de cada caso.

Los scritps son iguales que en el módulo anterior.

### CODE REVIEW

**Review by gscharly — COMMENTED** [review]
Hola @Felipe0542 

- En esta PR se te están volviendo a colar los cambios de los módulos pasados. En la rama main no tienes los cambios tampoco. ¿Cómo estás haciendo los mergeos?

- A la hora de hacer optimización de hiperparámetros, muestras en una gráfica un montón de curvas y colores, y resulta bastante difícil decidirse por una combinación. Hay representaciones mejores, y cuando pruebas muchas combinaciones lo realmente importante es la métrica objetivo. Primero puedes seleccionar los mejores K conjuntos de hiperparámetros, donde K sea un número pequeño, y luego representar las curvas para tomar la decisión final.

- Veo que aplicas una normalización a los datos antes de entrenar el Random Forest. Esto no es estrictamente necesario, ya que los árboles son capaces de trabajar con variables sin escalar, al trabajar con buckets dentro de cada variable. Introducir escalado o no puede parecer intrascendente, pero lo cierto es que conforme se aumenta el tamaño del conjunto de datos, cualquier operación extra que no sea totalmente necesaria puede suponer una pérdida de valor.

- A la hora de elegir el mejor modelo, escoges el que tiene mayor número de árboles, aunque hay otras configuraciones con una métrica en validación parecida. Por lo general, siempre es mejor utilizar un modelo más simple (en este caso con menos árboles), y más cuando las métricas son iguales.

- Una posible razón por la que un GBT no mejore un RF es la mayor cantidad de hiperparámetros que posee, lo que suele hacer más difícil su entrenamiento. Además, la implementación de sklearn no es la mejor. Yo soy muy fan de [Lightgbm](https://lightgbm.readthedocs.io/en/stable/), pero la más mítica es [XGboost](https://xgboost.readthedocs.io/en/stable/). Otras como [Catboost](https://catboost.ai/) funcionan muy bien con variables categóricas.

- Al trabajar con algoritmos con tantos hiperparámetros, es difícil hacer una optimización manual. Existen multitud de herramientas que te pueden ayudar en el proceso, desde las más básicas de [sklearn](https://scikit-learn.org/stable/modules/grid_search.html) (cuidado con splits temporales...) hasta otras más sofisticadas basadas en [estadística bayesiana](https://github.com/bayesian-optimization/BayesianOptimization)

- Por lo general, en problemas tabulares las redes neuronales siempre suelen salir perdiendo contra los modelos basados en árboles. Este [post](https://medium.com/geekculture/why-tree-based-models-beat-deep-learning-on-tabular-data-fcad692b1456) (que resume este [paper](https://proceedings.neurips.cc/paper_files/paper/2022/file/0378c7692da36807bdec87ab043cdadc-Paper-Datasets_and_Benchmarks.pdf)) intenta explicarlo de forma empírica

- A la hora de introducir variables categóricas, te recomiendo probar Catboost (mencionado antes) y [TargetEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.TargetEncoder.html) siempre por delante de OHE u OrdinalEncoder.

- Me falta la parte de interpretabilidad del modelo final elegido. ¿Qué variables son más importantes y cómo impactan al modelo? Si trabajas con árboles, suelo utilizar una librería llamada [SHAP](https://shap.readthedocs.io/en/latest/) que proporciona unos gráficos muy explicativos acerca de las variables que utiliza un modelo.
